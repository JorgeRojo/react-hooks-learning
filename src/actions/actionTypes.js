// COUNTER
export const COUNTER_DECREASE = 'COUNTER_DECREASE';
export const COUNTER_INCREASE = 'COUNTER_INCREASE';
export const COUNTER_RESET = 'COUNTER_RESET';
export const COUNTER_SET = 'COUNTER_SET';
// HTTP_REQUEST
export const HTTP_REQUEST = 'HTTP_REQUEST';
// API
export const API = 'API';
export const API_REQUEST = 'API/REQUEST';
export const API_ERROR = 'API/ERROR';
export const API_SUCCESS = 'API/SUCCESS';
export const API_RESET = 'API/RESET';
