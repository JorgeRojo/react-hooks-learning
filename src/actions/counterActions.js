import {
  COUNTER_DECREASE,
  COUNTER_INCREASE,
  COUNTER_RESET,
  COUNTER_SET,
} from './actionTypes';

export const counterDecrease = () => ({
  type: COUNTER_DECREASE,
});

export const counterIncrease = () => ({
  type: COUNTER_INCREASE,
});

export const counterReset = () => ({
  type: COUNTER_RESET,
});

export const counterSet = number => ({
  type: COUNTER_SET,
  payload: { number },
});
