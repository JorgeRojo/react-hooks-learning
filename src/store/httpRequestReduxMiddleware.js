import { HTTP_REQUEST } from '../actions/actionTypes';

export class HttpRequest {
  _url = '';
  _body = {};
  headers = { 'Content-Type': 'application/json; charset=UTF-8' };
  method = 'GET';
  cache = 'default';
  mode = 'cors';

  constructor(url = '', params = {}, bodyData) {
    this.setURL(url, params);
    this._body = JSON.stringify(bodyData);
  }

  setURL(url, params) {
    const urlParams = Object.keys(params).map(
      key => `${key}=${encodeURI(params[key])}`
    );
    const queryParamsSplitter = url.includes('?') ? '&' : '?';
    this._url = urlParams.length
      ? `${url}${queryParamsSplitter}${urlParams.join('&')}`
      : url;
  }

  get url() {
    return this._url;
  }
  get body() {
    return this._body;
  }
}

const requestAction = nextActionType => {
  return {
    type: `${nextActionType}/REQUEST`,
  };
};

const successAction = (nextActionType, data, onRequestSuccess) => {
  if (typeof onRequestSuccess === 'function') {
    onRequestSuccess(data);
  }

  return {
    type: `${nextActionType}/SUCCESS`,
    payload: { response: { data } },
  };
};

const errorAction = (nextActionType, data, onRequestError) => {
  if (typeof onRequestError === 'function') {
    onRequestError(data);
  }

  return {
    type: `${nextActionType}/ERROR`,
    payload: data,
  };
};

const httpRequestReduxMiddleware = ({ dispatch }) => next => async action => {
  const {
    type,
    nextActionType,
    request,
    responseStatus,
    onRequestError,
    onRequestSuccess,
  } = action;
  if (type === HTTP_REQUEST) {
    let response;
    let data;

    dispatch(requestAction(nextActionType));

    try {
      response = await fetch(request.url, request);
    } catch (error) {
      return next(errorAction(nextActionType, error, onRequestError));
    }

    if (response.status === responseStatus) {
      try {
        data = await response.json();
      } catch (error) {
        return next(errorAction(nextActionType, error, onRequestError));
      }
    } else {
      return next(
        errorAction(
          nextActionType,
          { error: { unexpectedStatus: response.status } },
          onRequestError
        )
      );
    }

    return next(successAction(nextActionType, data, onRequestSuccess));
  }

  return next(action);
};

export default httpRequestReduxMiddleware;
