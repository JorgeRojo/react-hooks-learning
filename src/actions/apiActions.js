import { HTTP_REQUEST, API_RESET, API } from './actionTypes';
import { HttpRequest } from '../store/httpRequestReduxMiddleware';

export const apiGet = () => ({
  type: HTTP_REQUEST,
  nextActionType: API,
  request: new HttpRequest('https://jsonplaceholder.typicode.com/posts'),
  responseStatus: 200,
});

export const apiReset = () => ({
  type: API_RESET,
});
