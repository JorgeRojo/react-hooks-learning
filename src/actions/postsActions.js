import { HTTP_REQUEST, POSTS_RESET, POSTS } from './actionTypes';
import { HttpRequest } from '../store/httpRequestReduxMiddleware';

export const postsGet = () => ({
  type: HTTP_REQUEST,
  nextActionType: POSTS,
  request: new HttpRequest('https://jsonplaceholder.typicode.com/posts'),
  responseStatus: 200,
});

export const postsReset = () => ({
  type: POSTS_RESET,
});
