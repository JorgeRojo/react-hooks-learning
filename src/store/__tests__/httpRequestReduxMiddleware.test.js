import httpRequestReduxMiddleware, {
  HttpRequest,
} from '../httpRequestReduxMiddleware';
import { HTTP_REQUEST } from '../../actions/actionTypes';

describe('HttpRequest', () => {
  it('with default data', () => {
    const url = 'http://url.com?p0=v0';
    const params = {
      p1: 'v1',
      p2: 'v2',
    };
    const bodyData = { foo: 'bar' };
    const httpRequest = new HttpRequest(url, params, bodyData);

    expect(httpRequest).toMatchSnapshot();
    expect(httpRequest.url).toEqual(`${url}&p1=v1&p2=v2`);
    expect(httpRequest.body).toEqual(JSON.stringify(bodyData));
  });

  it('with url without params', () => {
    const url = 'http://url.com';
    const httpRequest = new HttpRequest(url);
    expect(httpRequest.url).toEqual(url);
  });

  it('with url without body', () => {
    const httpRequest = new HttpRequest();
    expect(httpRequest.body).toEqual(undefined);
  });

  it('with url without previous params', () => {
    const url = 'http://url.com';
    const params = {
      p1: 'v1',
      p2: 'v2',
    };
    const httpRequest = new HttpRequest(url, params);
    expect(httpRequest.url).toEqual(`${url}?p1=v1&p2=v2`);
  });
});

describe('httpRequestReduxMiddleware', () => {
  const nextActionType = 'NEX_ACTION';
  const request = new HttpRequest('http://url.com');
  const responseStatus = 200;
  const responseData = { foo: 'bar' };
  const response = {
    status: responseStatus,
    json: () => Promise.resolve(responseData),
  };

  let spyFetch;
  let store = {};
  let next;
  let onRequestError;
  let onRequestSuccess;

  beforeEach(() => {
    spyFetch = jest
      .spyOn(global, 'fetch')
      .mockImplementation(() => Promise.resolve(response));
    store.dispatch = jest.fn();
    next = jest.fn();
    onRequestError = jest.fn();
    onRequestSuccess = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('with unknown action type', async () => {
    const action = {
      type: 'UNKNOWN_ACTION',
    };
    await httpRequestReduxMiddleware(store)(next)(action);
    expect(store.dispatch).not.toHaveBeenCalled();
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith(action);
  });

  it('with success data', async () => {
    const action = {
      type: HTTP_REQUEST,
      nextActionType,
      request,
      responseStatus,
      onRequestError,
      onRequestSuccess,
    };

    await httpRequestReduxMiddleware(store)(next)(action);

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith({
      type: `${nextActionType}/REQUEST`,
    });

    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith({
      type: `${nextActionType}/SUCCESS`,
      payload: { response: { data: responseData } },
    });

    expect(onRequestError).not.toHaveBeenCalled();
    expect(onRequestSuccess).toHaveBeenCalledTimes(1);
    expect(onRequestSuccess).toHaveBeenCalledWith(responseData);
  });

  it('with success data but without success callback', async () => {
    const action = {
      type: HTTP_REQUEST,
      nextActionType,
      request,
      responseStatus,
    };

    await httpRequestReduxMiddleware(store)(next)(action);

    expect(onRequestSuccess).not.toHaveBeenCalled();
  });

  it('with error fetch call', async () => {
    const fetchError = { error: 'error' };
    spyFetch.mockImplementation(() => Promise.reject(fetchError));
    const action = {
      type: HTTP_REQUEST,
      nextActionType,
      request,
      responseStatus,
      onRequestError,
      onRequestSuccess,
    };

    await httpRequestReduxMiddleware(store)(next)(action);

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith({
      type: `${nextActionType}/REQUEST`,
    });

    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith({
      type: `${nextActionType}/ERROR`,
      payload: fetchError,
    });

    expect(onRequestSuccess).not.toHaveBeenCalled();
    expect(onRequestError).toHaveBeenCalledTimes(1);
    expect(onRequestError).toHaveBeenCalledWith(fetchError);
  });

  it('with wrong status code', async () => {
    const status = 201;
    spyFetch.mockImplementation(() => Promise.resolve({ ...response, status }));

    const action = {
      type: HTTP_REQUEST,
      nextActionType,
      request,
      responseStatus,
    };

    await httpRequestReduxMiddleware(store)(next)(action);

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith({
      type: `${nextActionType}/REQUEST`,
    });

    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith({
      type: `${nextActionType}/ERROR`,
      payload: { error: { unexpectedStatus: status } },
    });
  });

  it('with wrong json response function', async () => {
    const jsonError = { error: 'error' };
    spyFetch.mockImplementation(() =>
      Promise.resolve({
        ...response,
        json: () => Promise.reject(jsonError),
      })
    );

    const action = {
      type: HTTP_REQUEST,
      nextActionType,
      request,
      responseStatus,
    };

    await httpRequestReduxMiddleware(store)(next)(action);

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith({
      type: `${nextActionType}/REQUEST`,
    });

    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith({
      type: `${nextActionType}/ERROR`,
      payload: jsonError,
    });
  });
});
