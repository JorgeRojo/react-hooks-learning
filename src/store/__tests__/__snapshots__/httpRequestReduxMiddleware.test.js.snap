// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpRequest with default data 1`] = `
HttpRequest {
  "_body": "{\\"foo\\":\\"bar\\"}",
  "_url": "http://url.com?p0=v0&p1=v1&p2=v2",
  "cache": "default",
  "headers": Object {
    "Content-Type": "application/json; charset=UTF-8",
  },
  "method": "GET",
  "mode": "cors",
}
`;
